core/trino-spi/src/main/java/io/trino/spi/block/AbstractVariableWidthBlock.java

```diff
@@ -120,6 +120,23 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
         blockBuilder.writeBytes(getRawSlice(position), getPositionOffset(position) + offset, length);
     }
 
+    public void writeBytesTo(int[] positions, int positionsCount, BatchByteWriter byteWriter)
```

Comment:
Could this API be
`void copyPositions(int[] positions, int offset, int length, byte[] destination, int destinationOffset)` ?
You can also look into using similar logic as `copyPositions`, it would just copy into provided bytes array from given target offset instead of a Slice.

Reply:
Aren't you already doing an extra loop in `process...Positions` to figure out size of destination array ? You can potentially just do that before this method to size destination bytes array correctly. Or you can send in an output buffer class instead of raw bytes array which will apply resizing logic of ensureBytesCapacity when needed.
Can we make the API `void copyPositions(int[] positions, int srcIndex, int length, DynamicSliceOutput output, int[] positionOffsets, boolean[] isNull, int destIndex)` ? This might enable us to extract what we need without requiring multiple passes.

Reply:
> Aren't you already doing an extra loop in process...Positions to figure out size of destination array ? You can potentially just do that before this method to size destination bytes array correctly.

Yes but this loop does not calculate lengths, it uses it to update offsets and other appender state.

> Can we make the API void copyPositions(int[] positions, int srcIndex, int length, DynamicSliceOutput output, int[] positionOffsets, boolean[] isNull, int destIndex) ? This might enable us to extract what we need without requiring multiple passes.

so this method would also update `positionOffsets` and `isNull`? why is that better than the original API?

Reply:
> so this method would also update positionOffsets and isNull?

Yes, it would basically make this one method make all the necessary updates.
If it works, it would be better as it would avoid introducing `BatchByteWriter`. It could avoid the need for multiple passes, one to find non-nulls and another over the non-nulls to find offset and length. It's also easier to read as it would consolidate most of the logic in one place instead of having to hop between between caller of this method and the logic inside this method.